src/View/Helper/FormHelper.php

```diff
@@ -2672,4 +2675,29 @@ public function implementedEvents()
     {
         return [];
     }
+
+    public function getValuesSources()
+    {
+        return $this->_valuesSources;
+    }
+
+    public function setValuesSources($sources)
+    {
+        $this->_valuesSources = array_values(array_intersect((array)$sources, ['context', 'data', 'query']));
+
+        return $this;
+    }
+
+    public function getSourceValue($fieldname)
+    {
+        foreach ($this->getValuesSources() as $valuesSource) {
+            if ($valuesSource === 'context') {
+                return $this->_getContext()->val($fieldname);
+            }
+            if (isset($this->request->{$valuesSource}[$fieldname])) {
```

Comment:
> 1176:      * When reading values you will get `null` for keys/values that do not exist.
http://api.cakephp.org/3.3/source-class-Cake.Network.Request.html#1166-1203

Due to `data()` being a getter and setter I cannot even patch it to take a flag to be strict and through exceptions when values are not being found.